Support the pretraining samples in the legacy dataset conversion utility

RobotSail opened the pull request:
<!-- Thank you for contributing to InstructLab! -->

<!-- STEPS TO FOLLOW:
  1. Add a description of the changes (frequently the same as the commit description)
  2. Enter the issue number next to "Resolves #" below (if there is no tracking issue resolved, **remove that section**)
  3. Follow the steps in the checklist below, starting with the **Commit Message Formatting**.
-->

<!-- Uncomment this section with the issue number if an issue is being resolved
**Issue resolved by this Pull Request:**
Resolves #
--->

PR #1644 introduced a dataset conversion utility which provided the ability for datasets
in the HuggingFaces messages format to be converted into the legacy format used by Linux train.

However; SDG is now planning on providing a pretraining dataset which is represented as
phase007. This new dataset will usually only contain only a single message which can be identified
by the `pretraining` role inside the `messages` list. E.g.:

```json
{
	"messages": [
		{
			"role": "pretraining",
			"content": "<|user|>Why did the chicken cross the road?<|assistant|>To get to the other side!"
		}
	]
}
```

This PR provides support for this new dataset format by identifying if a dataset contains the
`pretraining` sample and then parsing it out to map into the legacy sample format:
```json
{
	"user": "Why did the chicken cross the road?",
	"assistant": "To get to the other side!",
	"system": ""
}
```


**Checklist:**

- [x] **Commit Message Formatting**: Commit titles and messages follow guidelines in the
  [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#summary).
- [x] [Changelog](https://github.com/instructlab/instructlab/blob/main/CHANGELOG.md) updated with breaking and/or notable changes for the next minor release.
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.

leseb commented on CHANGELOG.md:
```diff
@@ -44,6 +44,9 @@
   values as the `serve` section of the config.
+* Legacy Linux training now supports the new messages format. When a dataset is provided in the
+  HuggingFace messages format, `ilab` will automatically convert it back into the legacy format.
+* Legacy Linux training is now compatible with the phase007 pretraining format.
```
phase007? Where can I read more about that? :)

RobotSail replied:
We don't have any documentation up for phase007 yet as I believe this is still something being worked on by SDG and training. But you can read about how it's used by the full training library in https://github.com/instructlab/training/pull/137

leseb reviewed (commented):
Do we have existing tests that validate this? Do they need to be expanded? Thanks!

RobotSail commented:
> Do we have existing tests that validate this? Do they need to be expanded? Thanks!

I'll add tests for this.

RobotSail commented:
@leseb I've updated this PR to include a more comprehensive test suite. I've also moved some of the old tests that were trying to test the dataset conversion in the linux test suite into the `test_utils.py` file. It turns out that the way we were mocking the linux train function meant that it received no coverage, so these tests weren't actually doing anything.

leseb commented on tests/test_utils.py:
```diff
@@ -29,6 +31,157 @@ def test_validate_documents(self, git_clone_checkout):
+                        "role": "system",
+                        "content": "You are a friendly assistant",
+                    },
+                    {"role": "user", "content": "Who is pickle rick?"},
```
Love it!

booxter commented on src/instructlab/utils.py:
```diff
@@ -556,15 +566,89 @@ def split_hostport(hostport: str) -> tuple[str, int]:
+    Raises ValueError if neither the '<|user|>' or '<|assistant|>' tokens exist
+    within the given sample.
+    """
+    if SDGTokens.ASSISTANT not in pretraining:
```
nit:

```
for token in (ASSISTANT, USER):
    if token not in pretraining:
         raise ...
```

RobotSail replied:
That's a bit cleaner, I'll do this instead.

booxter commented on src/instructlab/utils.py:
```diff
@@ -556,15 +566,89 @@ def split_hostport(hostport: str) -> tuple[str, int]:
+        raise ValueError(
+            f"pretraining sample doesn't contain the '{SDGTokens.USER}' token: {pretraining}"
+        )
+    user, assistant = pretraining.split(SDGTokens.ASSISTANT)
```
Question: Is it ever possible that the `pretraining` contains multiple tokens? Wonder if this should be handled somehow. (`split` will then return 3+-tuple, which will fail to unpack the return value here.)

RobotSail replied:
For now no, but in the future if this is something we decide to add then we'll have to support it. I'll add a comment to the docstring indicating that the assumption is a single instance of each, and that user precedes assistant.

booxter commented on src/instructlab/utils.py:
```diff
@@ -556,15 +566,89 @@ def split_hostport(hostport: str) -> tuple[str, int]:
+    Given a pretraining message sample which contains the tokens '<|user|>' and '<|assistant|>',
+    return the strings corresponding to the user & assistant inputs.
+
+    This function assumes that a given message begins with the '<|user|>' token
```
Is this enforced somewhere in the format definition? Or is just implementation detail that may change later? (Could you please share a link to where the format is defined? Maybe add a link to the comment?)

RobotSail replied:
It's not necessarily enforced, but there's no logical reason for anything to exist before `<|user|>` since we're providing that as a separate message.

We can change this to instead extract all of the text between `<|user|>` and `<|assistant|>` rather than assuming that `<|user|>` is the very first token of the string.

booxter commented on src/instructlab/utils.py:
```diff
@@ -556,15 +566,89 @@ def split_hostport(hostport: str) -> tuple[str, int]:
+            (msg["content"] for msg in sample["messages"] if msg["role"] == "system"),
+            "",
+        )
+        if not pretraining:
```
nit: I think you could raise it before `system` is calculated.

RobotSail replied:
Good point.

booxter commented on src/instructlab/utils.py:
```diff
@@ -601,6 +701,10 @@ def ensure_legacy_dataset(
     Given a dataset that's either in the HF messages format or the legacy ilab train format,
     ensure that the returned dataset is always in the legacy ilab train format.
     """
+    if len(dataset) == 0:
```
nit:

```
if not dataset:
    return []
```

RobotSail replied:
I'll make this change.

booxter commented on tests/test_utils.py:
```diff
@@ -29,6 +31,157 @@ def test_validate_documents(self, git_clone_checkout):
+            }
+        ]
+        if exception:
+            with pytest.raises(ValueError, match=match):
```
nit: `raises(exception, ...)`?

RobotSail replied:
PyTest doesn't like this unfortunately. It requires you to provide the exact exception or else it'll throw some nasty errors.

RobotSail replied:
MyPy* not PyTest

booxter commented on tests/test_lab_train.py:
```diff
@@ -210,86 +209,6 @@ def test_invalid_data_dir(self, cli_runner: CliRunner):
             assert "Could not read from data directory" in result.output
             assert result.exit_code == 1
 
-    @patch("instructlab.utils.is_macos_with_m_chip", return_value=False)
```
Nice!

booxter reviewed (approved):
Assuming this reflects the format (I can't be sure since it's not referred from the patch it seems), the code does what it should do. I also love how testing is switched to target conversion functions instead of running through `click` runner.

danmcp commented:
@mergifyio rebase

danmcp commented:
@mergifyio rebase
